nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java

```diff
@@ -0,0 +1,254 @@
+        }
+        return results;
+
+    }
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) {
+        int maxTransformsToCache = context.getProperty(TRANSFORM_CACHE_SIZE).asInteger();
+        transformCache = Caffeine.newBuilder()
+                .maximumSize(maxTransformsToCache)
+                .build();
+        // Precompile the transform if it hasn't been done already (and if there is no Expression Language present)
+        if (!context.getProperty(JSLT_TRANSFORM).isExpressionLanguagePresent()) {
+            final String transform = context.getProperty(JSLT_TRANSFORM).getValue();
+            try {
+                transformCache.put(transform, Parser.compileString(transform));
+            } catch (JsltException je) {
+                throw new ProcessException("Error compiling JSLT transform: " + je.getMessage(), je);
+            }
+        }
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        final FlowFile original = session.get();
+        if (original == null) {
+            return;
+        }
+
+        final ComponentLog logger = getLogger();
+        final StopWatch stopWatch = new StopWatch(true);
+
+        JsonNode firstJsonNode;
+        try (final InputStream in = session.read(original)) {
+            firstJsonNode = readJson(in);
+        } catch (final Exception e) {
+            logger.error("Failed to transform {}; routing to failure", original, e);
+            session.transfer(original, REL_FAILURE);
+            return;
+        }
+
+        try {
+            final String transform = context.getProperty(JSLT_TRANSFORM).evaluateAttributeExpressions(original).getValue();
+            Expression jsltExpression = transformCache.get(transform, currString -> Parser.compileString(transform));
+
+            final JsonNode transformedJson = jsltExpression.apply(firstJsonNode);
+            final ObjectWriter writer = context.getProperty(PRETTY_PRINT).asBoolean() ? jsonObjectMapper.writerWithDefaultPrettyPrinter() : jsonObjectMapper.writer();
+            final Object outputObject;
+            if (transformedJson == null || transformedJson.isNull()) {
+                logger.warn("JSLT transform resulted in no data");
+                outputObject = null;
+            } else {
+                outputObject = transformedJson;
+            }
+            FlowFile transformed = session.write(original, out -> {
+                if (outputObject != null) {
+                    writer.writeValue(out, outputObject);
+                }
+            });
+            transformed = session.putAttribute(transformed, CoreAttributes.MIME_TYPE.key(), "application/json");
+            session.transfer(transformed, REL_SUCCESS);
+            session.getProvenanceReporter().modifyContent(transformed, "Modified With " + transform, stopWatch.getElapsed(TimeUnit.MILLISECONDS));
```

Comment:
Is this manual call to Provenance Reporter still necessary?

Reply:
There are a number of processors that do their own modifyContent in order to specify a particular message with more context. This one I copied from JoltTransformJSON, so I kept it for consistency.

Reply:
Thanks for the explanation @mattyb149!